### ISSUE TITLE
Usage with react-router 6 `preload`

### ISSUE THREAD

**Issue by insanecoding** [andreiduca/use-async-resource#1]
Hey Andrei and thanks for the awesome library! 
It's simple but powerful, I really like it.

I have a question regarding its usage with `react-router` 6.
Since [alpha.5](https://github.com/ReactTraining/react-router/releases/tag/v6.0.0-alpha.5), `react-router` supports the `preload` function.

It seems like the perfect use case for `use-async-resource`. The reader can be lazily initialized somewhere on top and then the api calls can be done in the `preload` function.

This strategy worked for me, I was able to fetch in parallel both the React components and the api resources.

But I get this stacktrace from react:

```
react-dom.development.js:73 Warning: Cannot update a component (`App`) while rendering a different 
component (`Routes`). To locate the bad setState() call inside `Routes`, follow the stack trace as 
described in https://fb.me/setstate-in-render
    in Routes (created by App)
    in Router (created by BrowserRouter)
    in BrowserRouter (created by App)
    in Suspense (created by App)
    in App
    in StrictMode
	
printWarning @ react-dom.development.js:73
error @ react-dom.development.js:45
warnAboutRenderPhaseUpdatesInDEV @ react-dom.development.js:24697
scheduleUpdateOnFiber @ react-dom.development.js:22495
dispatchAction @ react-dom.development.js:16316
eval @ useAsyncResource.js:20
eval @ app.js:26
eval @ index.js:43
E @ index.js:43
Routes @ index.js:39
renderWithHooks @ react-dom.development.js:15273
mountIndeterminateComponent @ react-dom.development.js:18330
beginWork @ react-dom.development.js:19791
beginWork$1 @ react-dom.development.js:24635
performUnitOfWork @ react-dom.development.js:23540
workLoopConcurrent @ react-dom.development.js:23525
renderRootConcurrent @ react-dom.development.js:23490
performConcurrentWorkOnRoot @ react-dom.development.js:22768
workLoop @ scheduler.development.js:597
flushWork @ scheduler.development.js:552
performWorkUntilDeadline @ scheduler.development.js:164
```

I'm wondering now if I'm doing something wrong. I tried for both `experimental` and `latest` (16.13.1) react versions, all the same.

Here is how I do it:
`App.js`:
```jsx
...
const Categories = React.lazy(() => import(/* webpackChunkName: "categories" */ './components/categories'));
...
const fetchCategories = () =>
  fetch('https://api.chucknorris.io/jokes/categories').then(r => r.json());

const App = () => {
  const [categoriesReader, getCategories] = useAsyncResource(fetchCategories);

  const preloadCategories = React.useCallback(() => {
    getCategories();
  }, []);

  return (
    <Suspense fallback={<Fallback />}>
      <Router>
        <Routes>
          {/*...*/}
          <Route
            path="categories"
            element={<Categories reader={categoriesReader} />}
            preload={preloadCategories}
          />
          {/*...*/}
        </Routes>
      </Router>
    </Suspense>
  );
```

and the component itself is:

```jsx
import React from 'react'

const Categories = ({ reader }) => {
  const data = reader();

  if (data === undefined) return null;

  return (
    <ul>
      {
        data.map(item => <li key={item}>{item}</li>)
      }
    </ul>
  )
}

export default Categories;
```

Can the warning be ignored? Or am I doing something wrong? 
Could you please take a look?

**Comment by andreiduca** [comment]
Hi @insanecoding .

I successfully reproduced this warning with a simple `useState` in the main App component:

```jsx
const App = () => {
  const [categories, setCategories] = useState();

  const preloadCategories = React.useCallback(() => {
    setCategories(['one', 'two', 'three']);
  }, []);

  return (
      <Router>
        <Routes>
          {/*...*/}
          <Route
            path="categories"
            element={<Categories list={categories} />}
            preload={preloadCategories}
          />
          {/*...*/}
        </Routes>
      </Router>
  );

const Categories = ({ list }) => {
  if (list === undefined) return null;

  return (
    <ul>
      {
        list.map(item => <li key={item}>{item}</li>)
      }
    </ul>
  )
}
```

From what I can tell, React Router 6 is calling the `preload` function synchronously with the rendering of the `element`, i.e. inside the `render` method that renders the `element`. React forbids that.

I suspect you can open this issue with the React Router team, as v6 is still experimental and it may be fixed until the beta or final release.

---

One other way to not get this warning in your specific case would be to manually defer the preloading function with a `setTimeout` of `0`:

```js
const preloadCategories = React.useCallback(() => {
  setTimeout(() => {
    getCategories();
  }, 0);
}, []);
```

But this will cause your `element` to render twice: once with an empty data reader, and immediately after with the data reader that throws the fetching Promise. So if your lazy component is already preloaded, you might end up seeing a flash of content: rendering of the empty state, immediately followed by the suspense fallback.

**Comment by insanecoding** [comment]
Hi @andreiduca , thanks for a quick reply!

I confirm, `setTimeout(..., 0)` fixes the problem but I understand the unwanted effects there.
It would be the best to let `react-router` folks know about the issue. I'll do it.

**Comment by insanecoding** [comment]
Hey @andreiduca , so it turns out that `setState` shouldn't be called inside of `preload` function, according to `react-router` folks:
https://github.com/ReactTraining/react-router/issues/7374#issuecomment-633617383

**Comment by andreiduca** [comment]
I understand the intent there, but there is no clear or official documentation regarding the `preload` attribute.

React Router v6 is still in alpha, with breaking changes appearing with any release. So until it's officially out, I cannot consider making updates to my custom hook.

**Comment by insanecoding** [comment]
Yes, I see. No worries, I just wanted to let you know. Let's see how `react-router` evolves.

**Comment by andreiduca** [comment]
Fixed in v2.1.0: exporting `dataReaderInitializer` as `preloadResource`.

Readme: https://github.com/andreiduca/use-async-resource#preloading-resources

--

@insanecoding: In your particular case, you can just use the new helper inside the Route component

```jsx
import { preloadResource } from 'use-async-resource';

//...
const App = () => {
  return (
    <Suspense fallback={<Fallback />}>
      <Router>
        <Routes>
          {/*...*/}
          <Route
            path="categories"
            element={<Categories />}
            preload={() => preloadResource(fetchCategories)}
          />
          {/*...*/}
        </Routes>
      </Router>
    </Suspense>
  );
}
```

Then just use the `useAsyncResource` hook inside the `Categories` component directly:

```jsx
const Categories = ({ reader }) => {
  const [categoriesReader] = useAsyncResource(fetchCategories, []);
  const data = categoriesReader();

  // no need for this anymore, as it's eagerly initialized
  // if (data === undefined) return null;

  return (
    <ul>
      {
        data.map(item => <li key={item}>{item}</li>)
      }
    </ul>
  )
}
```
